File: 백준/Silver/9461. 파도반 수열/파도반 수열.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class Main {

	public static void main(String[] args) throws IOException {

		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringBuilder sb = new StringBuilder();

		int t = Integer.parseInt(br.readLine());

		for (int i = 0; i < t; i++) {
			int n = Integer.parseInt(br.readLine());
			long result = bottomUp(n);
			sb.append(result).append('\n');
		}
		System.out.println(sb);
	}

	static long bottomUp(int n) {

		long[] memo = new long[n + 3];

		memo[0] = 1;
		memo[1] = 1;
		memo[2] = 1;
		memo[3] = 1;

		for (int i = 4; i <= n; i++) {
			memo[i] = memo[i - 2] + memo[i - 3];
		}

		return memo[n];
	}
}

File: 백준/Silver/9461. 파도반 수열/파도반 수열.py
import sys

n = int(sys.stdin.readline().rstrip())

for _ in range(n):
    m = int(sys.stdin.readline().rstrip())
    memo = [0 for _ in range(m + 1)]

    if m == 1:
        print(1)
        continue

    memo[0] = 0
    memo[1] = 1
    memo[2] = 1

    for i in range(3, m + 1):
        memo[i] = memo[i - 2] + memo[i - 3]

    print(memo[m])
